### ISSUE TITLE
Softmin

### ISSUE THREAD

**Issue by GilesBathgate** [calvinmccarter/idw-attention#1]
Hi, can `softmin()` work instead of negative log softmax, and if not any insights as to why? Thanks.

**Comment by calvinmccarter** [comment]
Good question! Softmin is equivalent to using negative distance (Eq 1 in the paper) with softmax. As Fig1 in the paper shows, this is hard to train, because when the key is very far or very close, the derivative is basically zero.

**Comment by GilesBathgate** [comment]
Thanks for your reply, I think I was getting confused.

```math
\text{softmin}(x)_i = \frac{e^{-x_i}}{\sum_{j} e^{-x_j}}
```

Then i substituted `-log(x)`

```math
\text{softmax}(x)_i = \frac{e^{\log(x_i)}}{\sum_{j} e^{\log(x_j)}} = \frac{x_i}{\sum_j x_j}
```

My thinking was that this is simpler to compute, but I think my formulation is wrong.

**Comment by GilesBathgate** [comment]
Should be:

```math
\text{softmax}(x)_i = \frac{e^{-\log(x_i)}}{\sum_{j} e^{-\log(x_j)}} = \frac{\frac{1}{x_i}}{\sum_{j} \frac{1}{x_j}}
```

Which is exactly what you show in the paper.
